golang/cosmos/x/swingset/keeper/extension_snapshotter.go

```diff
@@ -0,0 +1,270 @@
+package keeper
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/Agoric/agoric-sdk/golang/cosmos/x/swingset/types"
+	vstoragetypes "github.com/Agoric/agoric-sdk/golang/cosmos/x/vstorage/types"
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	snapshots "github.com/cosmos/cosmos-sdk/snapshots/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+)
+
+var _ snapshots.ExtensionSnapshotter = &SwingsetSnapshotter{}
+
+// SnapshotFormat 1 is a proto message containing an artifact name, and the binary artifact data
+const SnapshotFormat = 1
+
+type activeExtensionSnapshot struct {
+	// The block height of the snapshot in progress
+	height uint64
+	// The logger for this snapshot
+	logger log.Logger
+	// The function to retrieve the exported artifacts
+	retrieveSwingStoreExport func() error
+	// The writer for extensions payloads
+	payloadWriter snapshots.ExtensionPayloadWriter
+}
+
+// SwingsetSnapshotter manages Swingset snapshots, ensuring insensitivity to
+// sub-block timing of communication with the JS side, and enforcing
+// concurrency requirements.
+// The caller of this submodule must arrange block level commit synchronization,
+// to ensure the results are deterministic.
+type SwingsetSnapshotter struct {
+	isConfigured            func() bool
+	takeAppSnapshot         func(height int64)
+	newRestoreContext       func(height int64) sdk.Context
+	swingStoreSnapshotter   *SwingStoreSnapshotter
+	getSwingStoreExportData func(ctx sdk.Context) []*vstoragetypes.DataEntry
+	logger                  log.Logger
+	activeExtensionSnapshot *activeExtensionSnapshot
+}
+
+// NewSwingsetSnapshotter creates a new SwingsetSnapshotter
+func NewSwingsetSnapshotter(app *baseapp.BaseApp, swingStoreSnapshotter *SwingStoreSnapshotter, getSwingStoreExportData func(ctx sdk.Context) []*vstoragetypes.DataEntry) SwingsetSnapshotter {
+	return SwingsetSnapshotter{
+		isConfigured:    func() bool { return app.SnapshotManager() != nil },
```

Comment:
This seems like an inversion of responsibility... is there a way to refactor, or to somehow assert a connection between the returned value and this SwingsetSnapshotter?

Reply:
This `isConfigured` was introduced for testability in an earlier PR (this is a simple split refactor). It doesn't strictly need to be there.